### PR TITLE
Adding ParsedSubdomainMeshModifier and a test

### PR DESCRIPTION
**Pull request by snschune**
tagging @dschwen

### CODE REVIEW

**Comment by dschwen** [comment]
Looks nice. Surprisingly little code is needed to implement this.

File: test/tests/mesh_modifiers/parsed_subdomain/tests [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[Tests]
+  [./parsed_subdomain_mm]
+    type = 'Exodiff'
+    input = 'parsed_subdomain_mm.i'
+    exodiff = 'parsed_subdomain_mm_in.e'
+    cli_args = '--mesh-only'
```
Comment (dschwen) [inline test/tests/mesh_modifiers/parsed_subdomain/tests]: Looks like you need to disable recover for this test.

File: test/tests/mesh_modifiers/parsed_subdomain/parsed_subdomain_mm.i [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  ny = 10
+  xmax = 1
+  ymax = 1
+  uniform_refine = 2
+[]
+
+[MeshModifiers]
+  [./subdomains]
+    type = ParsedSubdomainMeshModifier
+    combinatorial_geometry = 'x>0.1&x<0.9&y>0.1&y<0.9'
```
Comment (jwpeterson) [inline test/tests/mesh_modifiers/parsed_subdomain/parsed_subdomain_mm.i]: We can have spaces in ParsedFunction expressions if we use quotes, right?  Might make this a bit easier to read...

File: framework/src/meshmodifiers/ParsedSubdomainMeshModifier.C [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  // Loop over the elements
+  for (MeshBase::element_iterator el = mesh.active_elements_begin(); el != mesh.active_elements_end(); ++el)
+  {
+    _func_params[0] = (*el)->centroid()(0);
+    _func_params[1] = (*el)->centroid()(1);
+    _func_params[2] = (*el)->centroid()(2);
+    bool contains = evaluate(_func_F);
+
+    if (contains && std::find(_excluded_ids.begin(), _excluded_ids.end(), (*el)->subdomain_id()) == _excluded_ids.end())
+      (*el)->subdomain_id() = _block_id;
```
Comment (permcody) [inline framework/src/meshmodifiers/ParsedSubdomainMeshModifier.C]: You've got a linear find embedded in an element loop here. I know this loop only runs once but a large function might overlay a large region of the domain running this part several times. You could easily sort the vector in the constructor and use binary search here.

**Comment by snschune** [comment]
So the way to go is make `_excluded_ids` a std::set and use `count`?

File: framework/src/meshmodifiers/ParsedSubdomainMeshModifier.C (hunk 3 above)
Comment (dschwen) [inline framework/src/meshmodifiers/ParsedSubdomainMeshModifier.C]: Or store the excluded IDs in a hash map?

**Comment by dschwen** [comment]
> So the way to go is make `_excluded_ids` a std::set and use `count`?
 Oh, didn't see this. Yes, this would probably be the best variant.

**Comment by permcody** [comment]
If the number of exclusions is small the hash map or set will probably be slower than the linear search over the existing vector. That's why I was thinking about just sorting it and using binary search. You can leave it if you want. I'll let you decide.

**Comment by snschune** [comment]
Is there a pre-made binary search function in the stdlib that works on std::vector? `bsearch` expects arrays, doesn't it? @dschwen is timing multiple options right now, I'll wait till he's reported back.

**Comment by dschwen** [comment]
Meh, it is not worth fussing about. 10 million set counts take ~40ms. Linear search in a vector of length 5 is barely slower.

**Comment by dschwen** [comment]
http://www.cplusplus.com/reference/algorithm/binary_search/

**Comment by dschwen** [comment]
Don't use `std::binary_search`` it is slower than set by more than a factor of two for sizes of 20.

**Comment by permcody** [comment]
With small vectors, binary search and linear search are on the same order.

However we are definitely in the land of over-optimizing here. @dschwen's result is interesting and I have indeed seen this before. It turns out that `binary_search` is usually slower than other algorithms like `lower_bound` or `equal_range` even on large sets and it's not intuitive why but there are reasons. I guess the take away is, just leave it :)

**Comment by snschune** [comment]
works for me.

**Comment by permcody** [comment]
👍

**Pull request merged**